[JBTM-3383] Improve ARJUNA016009: Caught::java.lang.NullPointerException log message

ivassile opened the pull request:
Issue: https://issues.redhat.com/browse/JBTM-3383

MAIN AS_TESTS QA_JTA
!QA_JTS_JDKORB !QA_JTS_OPENJDKORB !QA_JTS_JACORB !BLACKTIE !XTS !PERF NO_WIN !RTS !TOMCAT !JACOCO !LRA

ivassile commented:
Note: We need to review existing tests. After the change test [2] will fail, because it uses xid with globalTransactionId = null.

[2]
[ERROR] Failures:
[ERROR] XARecoveryModuleUnitTest.testXAResourceRecoveryHelperDeregisterLocking:278 Expecting to get the resource registered by helper because the first pass was run just now expected:<com.hp.mwtests.ts.jta.recovery.TestXAResource@3b2cf7ab> but was:<null>

ochaloup commented:
TESTIT

ochaloup commented on ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java:
```diff
@@ -736,33 +738,29 @@ private final void xaRecoveryFirstPass(NameScopedXAResource xares)
+                        Xid xid = it.next();
+                        if (xid != null) {
+                            byte[] globalTransactionId = xid.getGlobalTransactionId();
+                            if (globalTransactionId != null) {
```
A topic for discussion - would make sense to check for `null` the format and branch as well? Is there some experience that e.g. Oracle could not be filling the other parts of Xid and leaving them `null`?

ivassile replied:
I added check for branch == null.

ochaloup commented on ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java:
```diff
@@ -736,33 +738,29 @@ private final void xaRecoveryFirstPass(NameScopedXAResource xares)
+                            byte[] globalTransactionId = xid.getGlobalTransactionId();
+                            if (globalTransactionId != null) {
+                                if (jtaLogger.logger.isDebugEnabled()) {
+                                    jtaLogger.logger.debug("Recovered: "
```
A code style thing (not necessary to be addressed) - maybe the `debug` call message could be whole at single line.

ivassile replied:
Resolved

ochaloup commented on ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java:
```diff
@@ -1223,6 +1221,40 @@ private void saveContactedJndiName(final String jndiName) {
+            byte[] globalTransactionId = xid.getGlobalTransactionId();
+            byte[] branchQualifier = xid.getBranchQualifier();
+            stringBuilder.append("< ");
+            stringBuilder.append(xid.getFormatId());
```
If there is the check for `globalTransactionId` and `branchQualifier` in this printing, it could be good to check the format for `null` as well here.

ivassile replied:
xid.getFormatId() is int

ochaloup reviewed (changes requested):
I have few requests for adjustments but nothing crucial. The idea as whole seems good to me.

ochaloup reviewed (changes requested):
@ivassile I'm sorry for changing my opinion but I verified the XA spec and JTA spec and I think my consideration was wrong last time.
The spec does not require the global part of `Xid` has not to be null. Such Xid should not be removed from recovery.

Details from XA spec

```
Although "xa.h" constrains the length and byte alignment of the data element within an
XID, it does not specify the data’s contents. The only requirement is that both gtrid and
bqual, taken together, must be globally unique.
```
I would ask you for changing this PR in  this regard - ie. do not remove the null global id from the `trans` array. Please, change only the logging machinery to not throwing the NPE when it's null (either global, or branch, or format).

ivassile commented:
@ochaloup Thank you for the review. I'll analyze the new requirements and rework the PR.

mmusgrov commented:
> Note: We need to review existing tests. After the change test [2] will fail, because it uses xid with globalTransactionId = null.
> 
> [2]
> 
> ```
> [ERROR] Failures:
> [ERROR] XARecoveryModuleUnitTest.testXAResourceRecoveryHelperDeregisterLocking:278 Expecting to get the resource registered by helper because the first pass was run just now expected:<com.hp.mwtests.ts.jta.recovery.TestXAResource@3b2cf7ab> but was:<null>
> ```

Is this something you need us to help you with?

ivassile commented on ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java:
```diff
@@ -736,33 +738,29 @@ private final void xaRecoveryFirstPass(NameScopedXAResource xares)
+                                }
+                            } else {
+                                it.remove();
+                                jtaLogger.logger.warn("Failed to recover: Returned global transaction identifier is null. "
```
Resolved

ivassile commented on ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java:
```diff
@@ -736,33 +738,29 @@ private final void xaRecoveryFirstPass(NameScopedXAResource xares)
+                            }   
+                        }                                  
+                    }
+                    trans = transList.toArray(new Xid[transList.size()]);
```
Resolved

ivassile commented:
@ochaloup I corrected most of the requested changes. Based on [1], I think we still need to reject and not process transactions where gtrid and bqual are both null, because the combine value won't be unique.
I also modified XidImple to handle gtrid null value.

@mmusgrov As we discussed, it will be great if you can help with creating a new test which will simulate Oracle(not instanceof XidImple) transaction to replicate the NPE before this change. If we decide to keep my current suggestion and remove transactions where gtrid and bqual are both null, we also need to modify the existing test testXAResourceRecoveryHelperDeregisterLocking.

[1] Although "xa.h" constrains the length and byte alignment of the data element within an
XID, it does not specify the data’s contents. The only requirement is that both gtrid and
bqual, taken together, must be globally unique.

ochaloup reviewed (commented):
@ivassile thanks for the enhancements. It seems fine to me.
Just we need to fix the failing test (the `CORE` one, the `AS_TESTS` failure is an intermittent on Narayana CI). And it would be good to create a new test for this behaviour as suggested.

mmusgrov commented:
> @ivassile I'm sorry for changing my opinion but I verified the XA spec and JTA spec and I think my consideration was wrong last time.
> The spec does not require the global part of `Xid` has not to be null. A Xid in such format should not be removed from recovery processing.
> 
> Details from XA spec (https://pubs.opengroup.org/onlinepubs/009680699/toc.pdf)
> 
> ```
> Although "xa.h" constrains the length and byte alignment of the data element within an
> XID, it does not specify the data’s contents. The only requirement is that both gtrid and
> bqual, taken together, must be globally unique.
> ```
> 
> I would ask you for changing this PR in this regard - ie. do not remove the `xares` when global id is `null` from the `trans` array. Please, change only the logging machinery to not throwing the NPE when it's null (either global, or branch, or format).
> (such change will simultaneously not cause failing the `XARecoveryModuleUnitTest.testXAResourceRecoveryHelperDeregisterLocking`)

@ochaloup is correct, the spec only requires that the Xid be unique. @ivassile can you change the code to only check for zero length bytes in the Xid and if it length is zero then log a warning if the formaId is not equal to -1 (and if it equals -1 then maybe log an info message). It might also be worthwhile logging an info message if either of the gtrid_length or bqual_length vaues are zero.

mmusgrov commented:
And I think we will need separate tests for `gtrid_length` being zero and `bqual_length` being zero and both being zero,

ivassile commented:
@mmusgrov Implemented the requested changes. Let me know if this is correct.

ochaloup commented on ArjunaJTA/jta/classes/com/arjuna/ats/jta/xa/XidImple.java:
```diff
@@ -186,15 +187,17 @@ public final void copy(Xid xid) {
 
-				System.arraycopy(gtx, 0, _theXid.data, 0, gtx.length);
+				if (gtxlength > 0) {
+		              System.arraycopy(gtx, 0, _theXid.data, 0, gtxlength);  
```
It seems a strange indent is used here.

ivassile replied:
Corrected

ochaloup reviewed (commented):
To me the changes seem fine. I would like just to mention that there is a strange indent formatting at some lines. That could be adjusted.

ivassile commented on ArjunaJTA/jta/classes/com/arjuna/ats/jta/logging/jtaI18NLogger.java:
```diff
@@ -600,6 +600,14 @@
 
+    @Message(id = 16149, value = "Returned global transaction identifier and branch qualifier are null but format id is not -1. {0}", format = MESSAGE_FORMAT)
+    @LogMessage(level = WARN)
+    public void warn_recovery_transaction_id_and_branch_qualifier_are_null_wrong_format_id(String arg0);
```
Corrected

ochaloup reviewed (approved):
@ivassile thanks. It seems fine to me.

...oh, just there was merged another PR meanwhile (ie. https://github.com/jbosstm/narayana/pull/1759) which touches (but only party) the same issue

mmusgrov commented:
> @ivassile thanks. It seems fine to me.
> 
> ...oh, just there was merged another PR meanwhile (ie. #1759) which touches (but only party) the same issue. A rebase would be needed.

I resolved the conflicts before I saw your comment.
This PR is now good to merge so I will do that now.

mmusgrov commented:
Ah git is saying it cannot automatically rebase so you will need to do what Ondra suggested, something like
- git pull pr/1747
- git pull --rebase upstream master # where upstream is our repo (upstream ssh://git@github.com/jbosstm/narayana.git)
- fix the the merge conflict (git add, git rebase --continue)
- push to your own topic branch (ivassile:JBTM-3383)
and then we can merge your PR

thanks

ivassile commented:
@mmusgrov The PR is now rebased and can be merged. This change overrides the change from https://github.com/jbosstm/narayana/pull/1759 and resolves more NPE issues in the code.

mmusgrov commented:
@ivassile Before we merge this one I have asked @ochaloup why we have two issues and two PRs for the same problem (and it is unclear to me why we need both fixes).

ochaloup commented:
It seems it's ok I'm merging it. If there is some more remarks we can fix it in a follow-up PR, I assume. Thanks @ivassile for the fix.

mmusgrov commented:
@ochaloup see my other comments on this

mmusgrov commented:
It is best not to ignore PR comments before merging

mmusgrov commented:
> @ivassile Before we merge this one I have asked @ochaloup why we have two issues and two PRs for the same problem (and it is unclear to me why we need both fixes).

This is not a good approach - we will end up with 3 PRs and 3 issues for an non functional issue so we need to get to the bottom of it to avoid it happening again.

mmusgrov commented:
> It seems it's ok I'm merging it. If there is some more remarks we can fix it in a follow-up PR, I assume. Thanks @ivassile for the fix.

It seems it's ok I'm merging it. If there is some more remarks we can fix it in a follow-up PR, I assume. Thanks @ivassile for the fix.
